docs/core/extensions/generic-host.md

```diff
@@ -136,8 +132,83 @@ The configuration created by `ConfigureAppConfiguration` is available in [HostBu
+[sigint]: https://en.wikipedia.org/wiki/Signal_(IPC)#SIGINT
+[sigquit]: https://en.wikipedia.org/wiki/Signal_(IPC)#SIGQUIT
+[sigterm]: https://en.wikipedia.org/wiki/Signal_(IPC)#SIGTERM
+
+Before .NET 6, there wasn't a way for .NET code to gracefully handle SIGTERM. To work around this limitation,
+`ConsoleLifetime` would subscribe to <xref:System.AppDomain.ProcessExit?displayProperty=fullName>. When `ProcessExit` was raised,
+`ConsoleLifetime` would signal the host to stop, and block the `ProcessExit` thread, waiting for the host to stop.
+This would allow for the clean up code in the application to run &mdash; for example, <xref:Microsoft.Extensions.Hosting.IHost.StopAsync%2A?displayProperty=nameWithType> and code after <xref:Microsoft.Extensions.Hosting.HostingAbstractionsHostExtensions.Run%2A?displayProperty=nameWithType> in the `Main` method.
+
+This caused other issues because SIGTERM wasn't the only way `ProcessExit` was raised. It is also raised by code
+in the application calling `Environment.Exit`. `Environment.Exit` isn't a graceful way of shutting down a process
+in the `Microsoft.Extensions.Hosting` app model. It raises the `ProcessExit` event and then exits the process. The end of the
+`Main` method doesn't get executed. Background and foreground threads are terminated, and `finally` blocks are *not* executed.
+
+Since `ConsoleLifetime` blocked `ProcessExit` waiting for the host to shutdown, this behavior led to [deadlocks][deadlocks]
+from `Environment.Exit` also blocking waiting for `ProcessExit` to return. Additionally, since the SIGTERM handling was attempting
+to gracefully shut down the process, `ConsoleLifetime` would set the <xref:System.Environment.ExitCode> to `0`, which [clobbered][clobbered] the user's exit code passed to `Environment.Exit`.
+
+[deadlocks]: https://github.com/dotnet/runtime/issues/50397
+[clobbered]: https://github.com/dotnet/runtime/issues/42224
+
+In .NET 6, [POSIX signals][POSIX signals] are supported and handled. This allows for `ConsoleLifetime` to handle SIGTERM gracefully, and no longer get involved when `Environment.Exit` is invoked.
+
+[POSIX signals]: https://github.com/dotnet/runtime/issues/50527
+
+> [!TIP]
+> For .NET 6+, `ConsoleLifetime` no longer has logic to handle scenario `Environment.Exit`. Apps that call `Environment.Exit`, and need to do clean up logic, can subscribe to `ProcessExit` themselves. Hosting will no longer attempt to gracefully stop the host in this scenario.
+
+If your application uses hosting, and you want to gracefully stop the host, you can call
+<xref:Microsoft.Extensions.Hosting.IHostApplicationLifetime.StopApplication%2A?displayProperty=nameWithType> instead of `Environment.Exit`.
+
+### Hosting shutdown process
+
+The following sequence diagram shows how the signals are handled internally in the hosting code. It isn't necessary for the majority of users to understand this process. But for developers that need a deep understanding, this may help them get started.
```

Comment:
```suggestion
The following sequence diagram shows how the signals are handled internally in the hosting code. It isn't necessary for most users to understand this process. But for developers that need a deep understanding, this may help you get started.
```